libbeat/autodiscover/template/config.go

```diff
@@ -18,17 +18,24 @@
 package template
 
 import (
+	"github.com/elastic/go-ucfg"
+
 	"github.com/elastic/beats/v7/libbeat/common"
 	"github.com/elastic/beats/v7/libbeat/common/bus"
 	"github.com/elastic/beats/v7/libbeat/conditions"
 	"github.com/elastic/beats/v7/libbeat/keystore"
 	"github.com/elastic/beats/v7/libbeat/logp"
-	"github.com/elastic/go-ucfg"
 )
 
-// Mapper maps config templates with conditions, if a match happens on a discover event
-// the given template will be used as config
-type Mapper []*ConditionMap
+// Mapper maps config templates with conditions in ConditionMaps, if a match happens on a discover event
+// the given template will be used as config.
+// Mapper also includes the global Keystore object at `keystore` and `kubernetesKeystoresRegistry`, which
+// holds Kubernetes keystores for known namespaces
+type Mapper struct {
+	ConditionMaps               []*ConditionMap
+	keystore                    keystore.Keystore
+	kubernetesKeystoresRegistry *keystore.KubernetesKeystoresRegistry
```

Comment:
generic package/type with concrete dependency on k8s. I'd vote for a "provider" interface.

Is it valid to configure both, the keystore and the registry, or will have one precedence. when resolving a key, which keystore has priority if both are used?

Reply:
👍  for the provider

Regarding setting them both, it is answered in the next comment.